internal/driver/custommetadata.go

```diff
@@ -0,0 +1,84 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+//
+// Copyright (C) 2022 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package driver
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/edgexfoundry/go-mod-core-contracts/v2/errors"
+	contract "github.com/edgexfoundry/go-mod-core-contracts/v2/models"
+)
+
+// setCustomMetadata will return a map containing the fields provided in the call to the function
+func (onvifClient *OnvifClient) setCustomMetadata(device contract.Device, data []byte) (contract.Device, errors.EdgeX) {
+	var dataObj contract.ProtocolProperties
+
+	err := json.Unmarshal(data, &dataObj)
+	if err != nil {
+		return device, errors.NewCommonEdgeX(errors.KindServerError, "failed to unmarshal the json request body", err)
+	}
+	if len(dataObj) == 0 {
+		return device, errors.NewCommonEdgeX(errors.KindServerError, "no data in PUT command", err)
+	}
+	for key, value := range dataObj {
+		trimValue := strings.TrimSpace(value)
+		trimKey := strings.TrimSpace(key)
+		if trimValue == "delete" || trimValue == "Delete" {
+			delete(device.Protocols[CustomMetadata], trimKey)
+			continue
+		}
+		if len(trimKey) == 0 {
+			continue
+		}
+		device.Protocols[CustomMetadata][trimKey] = trimValue // create or update a field in CustomMetadata
+	}
+
+	return device, nil
```

Comment:
this code may suffer from race conditions on the device object, but that may be a limitation of the sdk and need to be enhanced there or we refactor this whole service to add locking on the devices.